platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -119,8 +121,13 @@ - (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshot
     self.loading = true;
     
     __weak __typeof__(self) weakSelf = self;
+    // mbgl::Scheduler::GetCurrent() scheduler means "run callback on current (ie UI/main) thread"
+    // capture weakSelf to avoid retain cycle if callback is never called (ie snapshot cancelled)
     _snapshotCallback = std::make_unique<mbgl::Actor<mbgl::MapSnapshotter::Callback>>(*mbgl::Scheduler::GetCurrent(), [=](std::exception_ptr mbglError, mbgl::PremultipliedImage image, mbgl::MapSnapshotter::Attributions attributions, mbgl::MapSnapshotter::PointForFn pointForFn) {
         __typeof__(self) strongSelf = weakSelf;
+        // If self had died, _snapshotCallback would have been destroyed and this block would not be executed
+        assert(strongSelf);
```

Comment:
We tend to prefer to use the `NSAssert` family of assert macros. I think in this case we might need to use `NSCAssert` since the `NSAssert` macro uses self.

Reply:
I think it's important that we call the completion block in the case where `_snapshotCallback` is destroyed. Is that straightforward?

Reply:
I pushed a change with NSCAssert.

I thinking calling the iOS completion block on cancellation is outside the scope of the core code -- that is, I don't think you'd want the callback destructor to block waiting for a forced run of the completion callback (which might not be on the same thread the destructor is running from). To handle this in `MGLMapSnapshotter`, I guess you'd want to explicitly call the completion callback with an error value from `cancel` and also from...what is it, `dealloc`? I guess the short answer is that it's not terribly difficult but also not totally straightforward. I think we can call it outside the scope of this fix, although I agree that it's a better interface if the callback always either succeeds or errors. FWIW, the Android interface also allows you to cancel without receiving a callback.